### ISSUE TITLE
Add method buffer(Predicate<T>)

### ISSUE THREAD

**Issue by balamaci** [reactor/reactor-core#227]
Hi, I think it makes sense to add another method Publisher.buffer(Predicate<T>) to act like adding to a buffer while the predicate condition is true, while if it's false it would start another buffer. 
Right now it's available a buffer(Publisher<?> other, Supplier<C> bufferSupplier) but it's not expressive to implement a simple bufferWhile:

Flux<String> colors = Flux.fromArray(new String[]{"red", "green", "blue",
                "red", "yellow", "#", "green", "green"});

```
    colors.publish(p -> p.filter(val -> ! val.equals("#"))
                         .buffer(p.filter(val -> val.equals("#")), (Supplier<ArrayList<String>>) ArrayList::new)
            )
            .subscribe(list -> {
        String listCommaSeparated = String.join(",", list);
        log.info("List {}", listCommaSeparated);
    });
```

or as Stephane Maldini pointed in the gitter chanel this could be done:
        Flux<String> colors = Flux.fromArray(new String[]{"red", "green", "blue",
                "red", "yellow", "#", "green", "green"});

```
    colors.window()
          .concatMap(window -> window.takeUntil(val -> val.equals("#")).buffer())
          .subscribe(list -> {
                String listCommaSeparated = String.join(",", list);
                log.info("List {}", listCommaSeparated);
            });
```

**Comment by smaldini** [comment]
I think we can inline this in one operator as it is a common use case.

**Comment by simonbasle** [comment]
@smaldini do you mean a dedicated `FluxXXX` operator, or an alias? Quickly looking at this I'm thinking about `bufferUntil(Predicate)` and `bufferWhile(Predicate)`, to have a parallel to takeUntil/takeWhile... Not sure about the API surface though...
Did you also want to do the same for `window`?

**Comment by simonbasle** [comment]
and as noted in original comment, implementing this as an alias is quite simple:
```java
public final Flux<List<T>> bufferWhile(Predicate<? super T> predicate) {
	return onAssembly(this
		.window()
                                                             //or .takeUntil
		.concatMap(window -> window.takeWhile(predicate).buffer()));
	}
```

**Comment by simonbasle** [comment]
a dedicated operator is being implemented in #259, I'll split the `window` issue into a separate issue for `3.1.0`
